nuxeo-core/nuxeo-core-io/src/test/java/org/nuxeo/ecm/core/io/avro/TestAvroFactoryService.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * (C) Copyright 2018 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     pierre
+ */
+package org.nuxeo.ecm.core.io.avro;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+import java.util.Arrays;
+import java.util.Collection;
+
+import javax.inject.Inject;
+
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Field;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.nuxeo.ecm.core.schema.DocumentType;
+import org.nuxeo.ecm.core.schema.Namespace;
+import org.nuxeo.ecm.core.schema.SchemaManager;
+import org.nuxeo.ecm.core.schema.types.SchemaImpl;
+import org.nuxeo.ecm.core.schema.types.primitives.StringType;
+import org.nuxeo.runtime.avro.AvroFactoryContext;
+import org.nuxeo.runtime.avro.AvroFactoryService;
+import org.nuxeo.runtime.test.runner.Deploy;
+import org.nuxeo.runtime.test.runner.Features;
+import org.nuxeo.runtime.test.runner.FeaturesRunner;
+import org.nuxeo.runtime.test.runner.RuntimeFeature;
+
+/**
+ * @since 10.2
+ */
+@RunWith(FeaturesRunner.class)
+@Features(RuntimeFeature.class)
+@Deploy("org.nuxeo.ecm.core")
+@Deploy("org.nuxeo.ecm.core.io")
+@Deploy("org.nuxeo.ecm.core.schema")
+@Deploy("org.nuxeo.runtime.stream")
+public class TestAvroFactoryService {
+
+    private static Collection<String> FORBIDDEN = Arrays.asList("-", "__tutu__", "-gru__du__buk-");
+
+    private static final String FIELD = "field";
+
+    @Inject
+    public AvroFactoryService service;
+
+    @Inject
+    public SchemaManager schemaManager;
+
+    @Test
+    public void testDocumentTypeFactory() {
+        for (DocumentType documentType : schemaManager.getDocumentTypes()) {
+            AvroFactoryContext context = service.createContext();
+            Schema avro = context.createSchema(documentType);
+            assertNotNull(avro);
+            assertEquals(documentType.getSchemas().size(), avro.getFields().size());
+            avro.toString(true);
+        }
```

Comment:
cool, may be it is missing a unit test that use a complex custom document type and put an assertion on the expected json Avro schema, so by looking a the test I can see how is converted and make sure all the cases are covered

Reply:
this test is fine but don't give an idea if all cases are covered neither it gives an idea of the json avro schema corresponding to a document type which is the goal of this ticket. Adding a custom document type with complex and scalar fields and assert on the expected canonical schema will be instructive for any reader, a proof of task completed and a good place holder for further debug.

Reply:
Right, I'm gonna do that